erc1155/src/lib.rs

```diff
@@ -155,86 +148,115 @@ impl ERC1155TokenBase for ERC1155Token {
             panic!("ERC1155: ids and amounts length mismatch")
         }
 
-        for (i, ele) in ids.iter().enumerate() {
-            let amount = amounts[i];
-
-            let from_balance = self.balance_of(from, ele);
-
-            if from_balance < amount {
-                panic!("ERC1155: insufficient balance for transfer")
+        for (id, amount) in ids.iter().zip(amounts) {
+            if !self.can_transfer(from, id, *amount) {
+                panic!("ERC1155: all batch element should be transerfable");
             }
-
-            self.set_balance(from, ele, from_balance.saturating_sub(amount));
-            let to_balance = self.balance_of(to, ele);
-            self.set_balance(to, ele, to_balance.saturating_add(amount));
         }
+
+        ids.iter()
+            .enumerate()
+            .for_each(|(idx, x)| self.safe_transfer_from(from, to, x, amounts[idx]));
```

Comment:
We are either to use `zip` or rename the variables.
```suggestion
            .for_each(|(i, id)| self.safe_transfer_from(from, to, id, amounts[i]));
```